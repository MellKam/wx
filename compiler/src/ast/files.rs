use core::ops::Range;

use codespan_reporting::files;

#[derive(Debug, Clone)]
pub struct File {
    pub name: String,
    pub source: String,
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use core::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct FileId(u32);

#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    pub fn add(&mut self, name: impl Into<String>, source: impl Into<String>) -> Option<FileId> {
        use core::convert::TryFrom;

        let file_id = FileId(u32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    pub fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(&self, file_id: FileId, byte_index: usize) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(&self, file_id: FileId, line_index: usize) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}
