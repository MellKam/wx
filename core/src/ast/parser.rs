use super::{
    diagnostics::{Diagnostic, DiagnosticKind, DiagnosticStoreCell},
    lexer::{Lexer, LexerError, Token, TokenWithSpan},
};
use std::{cell::Cell, iter::Peekable, ops::Range};
use string_interner::{StringInterner, backend::StringBackend, symbol::SymbolU32};

#[derive(Debug, Clone, PartialEq)]
pub struct ExpressionId(usize);

#[derive(Debug, Clone, PartialEq)]
pub struct StatementId(usize);

#[derive(Debug, Clone, PartialEq)]
pub struct IntExpression {
    pub value: i64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct FloatExpression {
    pub value: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct StringExpression {
    pub value: SymbolU32,
}

#[derive(Debug, Clone, PartialEq)]
pub struct IdentifierExpression {
    pub value: SymbolU32,
}

#[derive(Debug, Clone, PartialEq)]
pub enum UnaryOperationKind {
    Minus,
    Not,
}

#[derive(Debug, Clone, PartialEq)]
pub struct UnaryOperator {
    pub kind: UnaryOperationKind,
    pub span: Range<usize>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct UnaryExpression {
    pub operator: UnaryOperator,
    pub operand: ExpressionId,
}

#[derive(Debug, Clone, PartialEq)]
pub enum BinaryOperationKind {
    // Arithmetic
    Plus,
    Minus,
    Multiply,
    Divide,
    Modulo,
    // Relational
    Equals,
    NotEquals,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
}

#[derive(Debug, Clone, PartialEq)]
pub struct BinaryOperator {
    pub kind: BinaryOperationKind,
    pub span: Range<usize>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct BinaryExpression {
    pub left: ExpressionId,
    pub right: ExpressionId,
    pub operator: BinaryOperator,
}

#[derive(Debug, Clone, PartialEq)]
pub struct AssignmentExpression {
    pub left: ExpressionId,
    pub right: ExpressionId,
}

#[derive(Debug, Clone, PartialEq)]
pub struct CallExpression {
    pub callee: ExpressionId,
    pub arguments: Vec<ExpressionId>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MemberExpression {
    pub object: ExpressionId,
    pub property: ExpressionId,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionKind {
    Int(IntExpression),
    Float(FloatExpression),
    String(StringExpression),
    Identifier(IdentifierExpression),
    Unary(UnaryExpression),
    Binary(BinaryExpression),
    Assignment(AssignmentExpression),
    Call(CallExpression),
    Member(MemberExpression),
}

#[derive(Debug, Clone, PartialEq)]
struct Expression {
    id: ExpressionId,
    kind: ExpressionKind,
}

#[derive(Debug, PartialEq)]
pub enum StatementKind {
    Expression(ExpressionId),
}

#[derive(Debug, PartialEq)]
pub struct Statement {
    id: StatementId,
    kind: StatementKind,
}

#[derive(Debug, PartialEq)]
struct Ast {
    expressions: Vec<Expression>,
    statements: Vec<Statement>,
}

impl Ast {
    pub fn new() -> Self {
        Self {
            expressions: Vec::new(),
            statements: Vec::new(),
        }
    }

    pub fn add_expression(&mut self, kind: ExpressionKind) -> ExpressionId {
        let id = ExpressionId(self.expressions.len());
        self.expressions.push(Expression {
            id: id.clone(),
            kind,
        });
        return id;
    }

    pub fn add_statement(&mut self, kind: StatementKind) -> StatementId {
        let id = StatementId(self.statements.len());
        self.statements.push(Statement {
            id: id.clone(),
            kind,
        });
        return id;
    }
}

#[derive(Debug, Clone, PartialEq, PartialOrd)]
pub enum BindingPower {
    Default,
    Assignment,
    Logical,
    Relational,
    Additive,
    Multiplicative,
    Call,
    Member,
    Primary,
}

type NudHandler = fn(parser: &mut Parser) -> Result<ExpressionId, String>;
type LedHandler =
    fn(parser: &mut Parser, left: ExpressionId, bp: BindingPower) -> Result<ExpressionId, String>;

struct Parser<'a> {
    interner: StringInterner<StringBackend>,
    diagnostics: DiagnosticStoreCell,
    source: &'a str,
    lexer: Peekable<Lexer<'a>>,
    ast: Ast,
}

enum ParserError {
    UnexpectedEndOfInput,
    NudHandlerNotFound,
}

impl<'a> Parser<'a> {
    pub fn parse() -> Result<Ast, ()> {
        todo!()
    }

    // fn peek(&mut self) -> Option<TokenWithSpan> {
    //     loop {
    //         let result = match self.lexer.peek() {
    //             Some(result) => result,
    //             None => return None,
    //         };

    //         match result {
    //             Ok(token) => return Some(token.clone()),
    //             Err(error) => {
    //                 match error.kind {
    //                     LexerError::UnexpectedCharacter => {
    //                         self.diagnostics.borrow_mut().diagnostics.push(Diagnostic {
    //                             message: "unexpected character".to_string(),
    //                             span: error.span.clone(),
    //                             kind: DiagnosticKind::Error,
    //                         });
    //                     }
    //                     _ => {

    //                     }
    //                 }

    //                 self.lexer.next();
    //             }

    fn parse_expression(
        &mut self,
        min_binding_power: BindingPower,
    ) -> Result<ExpressionId, String> {
        let token = match self.lexer.peek() {
            Some(Ok(token)) => token.clone(),
            _ => return Err("enexpected end of input".to_string()),
        };

        let nud_handler = match Parser::nud_lookup(token.kind) {
            Some(handler) => handler,
            None => return Err("nud handler not found".to_string()),
        };
        let mut left = nud_handler(self).unwrap();

        while let Some(Ok(token)) = self.lexer.peek() {
            let led_handler = match Parser::led_lookup(token.kind.clone()) {
                Some(handler) => handler,
                None => break,
            };

            let operator_binding_power = Parser::bp_lookup(token.kind.clone());
            if operator_binding_power <= min_binding_power {
                break;
            }

            left = led_handler(self, left, operator_binding_power).unwrap();
        }

        Ok(left)
    }

    fn nud_lookup(token: Token) -> Option<NudHandler> {
        match token {
            Token::Int => Some(parse_int_expression),
            Token::Float => Some(parse_float_expression),
            Token::Identifier => Some(parse_identifier_expression),
            _ => None,
        }
    }

    fn led_lookup(token: Token) -> Option<LedHandler> {
        todo!()
    }

    fn bp_lookup(token: Token) -> BindingPower {
        match token {
            Token::Plus | Token::Minus => BindingPower::Additive,
            Token::Star | Token::Slash => BindingPower::Multiplicative,
            Token::EqualEqual
            | Token::BangEqual
            | Token::LeftAngle
            | Token::LessEqual
            | Token::RightAngle
            | Token::GreaterEqual => BindingPower::Relational,
            Token::Equal => BindingPower::Assignment,
            Token::AmperAmper | Token::VbarVbar => BindingPower::Logical,
            Token::Dot => BindingPower::Member,
            Token::LeftParen => BindingPower::Call,
            _ => BindingPower::Default,
        }
    }
}

fn parse_int_expression(parser: &mut Parser) -> Result<ExpressionId, String> {
    let token = parser.lexer.next().unwrap().unwrap();
    if token.kind != Token::Int {
        return Err("expected integer".to_string());
    }

    let value = match parser.source[token.span.clone()].parse::<i64>() {
        Ok(value) => value,
        Err(_) => {
            parser
                .diagnostics
                .borrow_mut()
                .diagnostics
                .push(Diagnostic {
                    message: "Invalid integer literal".to_string(),
                    span: token.span.clone(),
                    kind: DiagnosticKind::Error,
                });

            // continue parsing with a dummy value
            0
        }
    };

    let expression_id = parser
        .ast
        .add_expression(ExpressionKind::Int(IntExpression { value }));
    Ok(expression_id)
}

fn parse_float_expression(parser: &mut Parser) -> Result<ExpressionId, String> {
    let token = parser.lexer.next().unwrap().unwrap();
    if token.kind != Token::Float {
        return Err("expected float".to_string());
    }

    let value = match parser.source[token.span.clone()].parse::<f64>() {
        Ok(value) => value,
        Err(_) => {
            parser
                .diagnostics
                .borrow_mut()
                .diagnostics
                .push(Diagnostic {
                    message: "Invalid float literal".to_string(),
                    span: token.span.clone(),
                    kind: DiagnosticKind::Error,
                });

            // continue parsing with a dummy value
            0.0
        }
    };

    let expression_id = parser
        .ast
        .add_expression(ExpressionKind::Float(FloatExpression { value }));
    Ok(expression_id)
}

fn parse_identifier_expression(parser: &mut Parser) -> Result<ExpressionId, String> {
    let token = parser.lexer.next().unwrap().unwrap();
    if token.kind != Token::Identifier {
        return Err("expected identifier".to_string());
    }

    let value = parser
        .interner
        .get_or_intern(&parser.source[token.span.clone()]);

    let expression_id = parser
        .ast
        .add_expression(ExpressionKind::Identifier(IdentifierExpression { value }));
    Ok(expression_id)
}

fn parse_grouping_expression(parser: &mut Parser) -> Result<ExpressionId, String> {
    let left_paren = parser.lexer.next().unwrap().unwrap();
    let expression_id = parser.parse_expression(BindingPower::Primary)?;
    let right_paren = parser.lexer.next().unwrap().unwrap();
    if right_paren.kind != Token::RightParen {
        return Err("expected right parenthesis".to_string());
    }

    Ok(expression_id)
}
