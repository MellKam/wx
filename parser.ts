import { Token, TokenKind } from "./lexer.ts";

export const ExpressionKind = {
	Int: 0, // 123
	Float: 1, // 123.45
	Identifier: 2, // x
	Unary: 3, // -1
	Binary: 4, // 1 + 2
	Assignment: 5, // x = 1
} as const;
export type ExpressionKind =
	(typeof ExpressionKind)[keyof typeof ExpressionKind];

export interface Expression {
	kind: ExpressionKind;
	start: number;
	end: number;
}

export const StatementKind = {
	Expression: 0, // 1 + 2;
	VariableDeclaration: 1, // let x: i32 = 1;
	FunctionDeclaration: 2, // fn add(x: i32, y: i32): i32 { return x + y; }
	FunctionReturn: 3, // return 1;
} as const;
export type StatementKind = (typeof StatementKind)[keyof typeof StatementKind];

export interface Statement {
	kind: StatementKind;
	start: number;
	end: number;
}

export const BindingPower = {
	Default: 0,
	Assignment: 1,
	Additive: 2,
	Multiplicative: 3,
	Primary: 4,
} as const;
export type BindingPower = (typeof BindingPower)[keyof typeof BindingPower];

export class TokenIterator {
	readonly tokens: Token[];
	position: number;

	constructor(tokens: Token[]) {
		this.tokens = tokens;
		this.position = 0;
	}

	current(): Token | null {
		return this.tokens[this.position] || null;
	}

	advance(): Token | null {
		const previous_token = this.tokens[this.position] || null;
		if (previous_token) this.position++;
		return previous_token;
	}
}

type NullDenotationHandler = (
	parser: Parser,
	tokens: TokenIterator
) => Expression;
type LeftDenotationHandler = (
	parser: Parser,
	tokens: TokenIterator,
	left: Expression,
	binding_power: BindingPower
) => Expression;
type StatementHandler = (parser: Parser, tokens: TokenIterator) => Statement;

export class Parser {
	readonly binding_power_lookup: Map<TokenKind, BindingPower>;
	readonly nud_lookup: Map<TokenKind, NullDenotationHandler>;
	readonly led_lookup: Map<TokenKind, LeftDenotationHandler>;
	readonly statement_lookup: Map<TokenKind, StatementHandler>;

	constructor(
		binding_power_lookup: Map<TokenKind, BindingPower>,
		nud_lookup: Map<TokenKind, NullDenotationHandler>,
		led_lookup: Map<TokenKind, LeftDenotationHandler>,
		statement_lookup: Map<TokenKind, StatementHandler>
	) {
		this.binding_power_lookup = binding_power_lookup;
		this.nud_lookup = nud_lookup;
		this.led_lookup = led_lookup;
		this.statement_lookup = statement_lookup;
	}

	static create(): Parser {
		const binding_power_lookup = new Map<TokenKind, BindingPower>();
		const nud_lookup = new Map<TokenKind, NullDenotationHandler>();
		const led_lookup = new Map<TokenKind, LeftDenotationHandler>();
		const statement_lookup = new Map<TokenKind, StatementHandler>();

		const leds: Array<
			[token: TokenKind, bp: BindingPower, handler: LeftDenotationHandler]
		> = [
			[TokenKind.Assign, BindingPower.Assignment, parse_assignment_expression],
			[TokenKind.Plus, BindingPower.Additive, parse_binary_expression],
			[TokenKind.Dash, BindingPower.Additive, parse_binary_expression],
			[TokenKind.Star, BindingPower.Multiplicative, parse_binary_expression],
			[TokenKind.Slash, BindingPower.Multiplicative, parse_binary_expression],
			[TokenKind.Percent, BindingPower.Multiplicative, parse_binary_expression],
		];
		for (const [token, bp, handler] of leds) {
			binding_power_lookup.set(token, bp);
			led_lookup.set(token, handler);
		}

		const nuds: Array<[token: TokenKind, handler: NullDenotationHandler]> = [
			[TokenKind.Identifier, parse_identifier_expression],
			[TokenKind.IntLiteral, parse_int_expression],
			[TokenKind.FloatLiteral, parse_float_expression],
			[TokenKind.OpenParen, parse_grouping_expression],
			[TokenKind.Dash, parse_unary_expression],
		];
		for (const [token, handler] of nuds) {
			binding_power_lookup.set(token, BindingPower.Primary);
			nud_lookup.set(token, handler);
		}

		const statements: Array<[token: TokenKind, handler: StatementHandler]> = [
			[TokenKind.LetKeyword, parse_variable_declaration_statement],
			[TokenKind.FnKeyword, parse_function_declaration_statement],
			[TokenKind.ReturnKeyword, parse_function_return_statement],
		];
		for (const [token, handler] of statements) {
			binding_power_lookup.set(token, BindingPower.Default);
			statement_lookup.set(token, handler);
		}

		return new Parser(
			binding_power_lookup,
			nud_lookup,
			led_lookup,
			statement_lookup
		);
	}

	parse(tokens: Token[]): Statement[] {
		const token_iterator = new TokenIterator(tokens);
		const body: Statement[] = [];

		while (true) {
			const token = token_iterator.current();
			if (!token) break;
			const statement_handler =
				this.statement_lookup.get(token.kind) || parse_expression_statement;

			const statement = statement_handler(this, token_iterator);
			body.push(statement);
		}

		return body;
	}

	parse_expression(tokens: Token[]): Expression {
		const token_iterator = new TokenIterator(tokens);
		return parse_expression(this, token_iterator, BindingPower.Default);
	}
}

const parse_expression = (
	parser: Parser,
	tokens: TokenIterator,
	min_binding_power: BindingPower
): Expression => {
	const token = tokens.current();
	if (!token) {
		throw new Error("No token found");
	}
	const nud_handler = parser.nud_lookup.get(token.kind);
	if (!nud_handler) {
		throw new Error(`No NUD handler found for token ${JSON.stringify(token)}`);
	}
	let left = nud_handler(parser, tokens);

	while (true) {
		const token = tokens.current();
		if (!token) break;
		const operator_binding_power = parser.binding_power_lookup.get(token.kind);
		if (
			operator_binding_power === undefined ||
			operator_binding_power <= min_binding_power
		) {
			break;
		}

		const led_handler = parser.led_lookup.get(token.kind);
		if (!led_handler) {
			throw new Error("No LED handler found");
		}

		left = led_handler(parser, tokens, left, operator_binding_power);
	}

	return left;
};

export interface IntExpression extends Expression {
	kind: typeof ExpressionKind.Int;
	value: number;
	start: number;
	end: number;
}

const parse_int_expression: NullDenotationHandler = (
	_parser,
	tokens
): IntExpression => {
	const token = tokens.advance();
	if (!token) throw new Error("No token found");
	const value = parseInt(token.value);
	if (isNaN(value)) throw new Error("Failed to parse integer");

	return {
		kind: ExpressionKind.Int,
		value,
		start: token.start,
		end: token.end,
	};
};

export interface FloatExpression extends Expression {
	kind: typeof ExpressionKind.Int;
	value: number;
	start: number;
	end: number;
}

const parse_float_expression: NullDenotationHandler = (
	_parser,
	tokens
): FloatExpression => {
	const token = tokens.advance();
	if (!token) throw new Error("No token found");
	const value = parseFloat(token.value);
	if (isNaN(value)) throw new Error("Failed to parse float");

	return {
		kind: ExpressionKind.Int,
		value,
		start: token.start,
		end: token.end,
	};
};

export interface IdentifierExpression extends Expression {
	kind: typeof ExpressionKind.Identifier;
	value: string;
	start: number;
	end: number;
}

export const parse_identifier_expression: NullDenotationHandler = (
	_parser,
	tokens
): IdentifierExpression => {
	const token = tokens.advance();
	if (!token) throw new Error("Expected identifier");

	return {
		kind: ExpressionKind.Identifier,
		value: token.value,
		start: token.start,
		end: token.end,
	};
};

export interface UnaryExpression extends Expression {
	kind: typeof ExpressionKind.Unary;
	operator: TokenKind;
	right: Expression;
	start: number;
	end: number;
}

const parse_unary_expression: NullDenotationHandler = (
	parser,
	tokens
): UnaryExpression => {
	const operator = tokens.advance();
	if (!operator) throw new Error("No token found");
	const right = parse_expression(parser, tokens, BindingPower.Primary);

	return {
		kind: ExpressionKind.Unary,
		operator: operator.kind,
		right,
		start: operator.start,
		end: right.end,
	};
};

const parse_grouping_expression: NullDenotationHandler = (parser, tokens) => {
	const _ = tokens.advance();
	const expression = parse_expression(parser, tokens, BindingPower.Default);
	const token = tokens.advance();
	if (!token) throw new Error("No token found");
	if (token.kind === TokenKind.CloseParen) {
		return expression;
	} else {
		throw new Error("Expected closing parenthesis");
	}
};

export interface BinaryExpression extends Expression {
	kind: typeof ExpressionKind.Binary;
	left: Expression;
	operator: TokenKind;
	right: Expression;
	start: number;
	end: number;
}

const parse_binary_expression: LeftDenotationHandler = (
	parser,
	tokens,
	left,
	binding_power
): BinaryExpression => {
	const operator = tokens.advance();
	if (!operator) throw new Error("No token found");
	const right = parse_expression(parser, tokens, binding_power);

	return {
		kind: ExpressionKind.Binary,
		left,
		operator: operator.kind,
		right,
		start: left.start,
		end: right.end,
	};
};

export interface VariableDeclarationStatement extends Statement {
	kind: typeof StatementKind.VariableDeclaration;
	identifier: string;
	type: string;
	value: Expression;
	start: number;
	end: number;
}

const parse_variable_declaration_statement: StatementHandler = (
	parser,
	tokens
): VariableDeclarationStatement => {
	const let_keyword = tokens.advance();
	if (!let_keyword || let_keyword.kind !== TokenKind.LetKeyword) {
		throw new Error("Expected let keyword");
	}

	const identifier = tokens.advance();
	if (!identifier || identifier.kind !== TokenKind.Identifier) {
		throw new Error("Expected identifier after let keyword");
	}

	const colon = tokens.advance();
	if (!colon || colon.kind !== TokenKind.Colon) {
		throw new Error("Expected colon after identifier");
	}

	const type = tokens.advance();
	if (!type || type.kind !== TokenKind.Identifier) {
		throw new Error("Expected type after colon");
	}

	const equals = tokens.advance();
	if (!equals || equals.kind !== TokenKind.Assign)
		throw new Error("Expected equal sign after identifier");

	const value = parse_expression(parser, tokens, BindingPower.Default);

	const semicolon = tokens.advance();
	if (!semicolon || semicolon.kind !== TokenKind.Semicolon) {
		throw new Error("Expected semicolon after variable declaration statement");
	}

	return {
		kind: StatementKind.VariableDeclaration,
		identifier: identifier.value,
		type: type.value,
		value,
		start: let_keyword.start,
		end: semicolon.end,
	};
};

export interface ExpressionStatement extends Statement {
	kind: typeof StatementKind.Expression;
	expression: Expression;
	start: number;
	end: number;
}

const parse_expression_statement: StatementHandler = (
	parser,
	tokens
): ExpressionStatement => {
	const expression = parse_expression(parser, tokens, BindingPower.Default);
	const semicolon = tokens.advance();
	if (!semicolon || semicolon.kind !== TokenKind.Semicolon) {
		throw new Error("Expected semicolon after expression statement");
	}

	return {
		kind: StatementKind.Expression,
		expression,
		start: expression.start,
		end: semicolon.end,
	};
};

export interface AssignmentExpression extends Expression {
	kind: typeof ExpressionKind.Assignment;
	identifier: string;
	value: Expression;
	start: number;
	end: number;
}

const parse_assignment_expression: LeftDenotationHandler = (
	parser,
	tokens,
	left,
	binding_power
): AssignmentExpression => {
	if (left.kind !== ExpressionKind.Identifier) {
		throw new Error("Expected identifier on the left side of assignment");
	}

	const equals = tokens.advance();
	if (!equals || equals.kind !== TokenKind.Assign) {
		throw new Error("Expected equal sign after identifier");
	}

	const value = parse_expression(parser, tokens, binding_power);

	return {
		kind: ExpressionKind.Assignment,
		identifier: (left as IdentifierExpression).value,
		value,
		start: left.start,
		end: value.end,
	};
};

export interface FunctionDeclarationStatement extends Statement {
	kind: typeof StatementKind.FunctionDeclaration;
	name: string;
	args: Array<[name: string, type: string]>;
	return_type: string;
	body: Statement[];
	start: number;
	end: number;
}

const parse_function_declaration_statement: StatementHandler = (
	parser,
	tokens
): FunctionDeclarationStatement => {
	const fn_keyword = tokens.advance();
	if (!fn_keyword || fn_keyword.kind !== TokenKind.FnKeyword) {
		throw new Error("Expected fn keyword");
	}

	const name = tokens.advance();
	if (!name || name.kind !== TokenKind.Identifier) {
		throw new Error("Expected identifier after fn keyword");
	}

	const open_paren = tokens.advance();
	if (!open_paren || open_paren.kind !== TokenKind.OpenParen) {
		throw new Error("Expected opening parenthesis after function name");
	}

	const args: [name: string, type: string][] = [];
	while (true) {
		const arg_name = tokens.advance();
		if (!arg_name || arg_name.kind !== TokenKind.Identifier) {
			throw new Error("Expected identifier as argument name");
		}

		const colon = tokens.advance();
		if (!colon || colon.kind !== TokenKind.Colon) {
			throw new Error("Expected colon after argument name");
		}

		const arg_type = tokens.advance();
		if (!arg_type || arg_type.kind !== TokenKind.Identifier) {
			throw new Error("Expected type after colon");
		}

		args.push([arg_name.value, arg_type.value]);

		const comma = tokens.advance();
		if (!comma) {
			throw new Error("Expected comma or closing parenthesis after argument");
		}
		if (comma.kind === TokenKind.CloseParen) {
			break;
		}
		if (comma.kind !== TokenKind.Comma) {
			throw new Error("Expected comma or closing parenthesis after argument");
		}
	}

	const colon = tokens.advance();
	if (!colon || colon.kind !== TokenKind.Colon) {
		throw new Error("Expected colon after closing parenthesis");
	}

	const return_type = tokens.advance();
	if (!return_type || return_type.kind !== TokenKind.Identifier) {
		throw new Error("Expected return type after colon");
	}

	const open_curly = tokens.advance();
	if (!open_curly || open_curly.kind !== TokenKind.OpenCurely) {
		throw new Error("Expected opening curly brace after return type");
	}

	const body: Statement[] = [];
	while (true) {
		const token = tokens.current();
		if (!token) break;
		if (token.kind === TokenKind.CloseCurely) break;

		const statement_handler =
			parser.statement_lookup.get(token.kind) || parse_expression_statement;
		const statement = statement_handler(parser, tokens);
		body.push(statement);
	}

	const close_curly = tokens.advance();
	if (!close_curly || close_curly.kind !== TokenKind.CloseCurely) {
		throw new Error("Expected closing curly brace after function body");
	}

	return {
		kind: StatementKind.FunctionDeclaration,
		name: name.value,
		args,
		return_type: return_type.value,
		body,
		start: fn_keyword.start,
		end: close_curly.end,
	};
};

export interface FunctionReturnStatement extends Statement {
	kind: typeof StatementKind.FunctionReturn;
	value: Expression;
	start: number;
	end: number;
}

const parse_function_return_statement: StatementHandler = (
	parser,
	tokens
): FunctionReturnStatement => {
	const return_keyword = tokens.advance();
	if (!return_keyword || return_keyword.kind !== TokenKind.ReturnKeyword) {
		throw new Error("Expected return keyword");
	}

	const value = parse_expression(parser, tokens, BindingPower.Default);

	const semicolon = tokens.advance();
	if (!semicolon || semicolon.kind !== TokenKind.Semicolon) {
		throw new Error("Expected semicolon after return statement");
	}

	return {
		kind: StatementKind.FunctionReturn,
		value,
		start: return_keyword.start,
		end: semicolon.end,
	};
};
