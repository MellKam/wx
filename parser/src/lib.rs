mod expression;
mod lexer;
mod statement;
mod unescape;

pub use expression::*;
pub use lexer::*;

pub fn add(left: u64, right: u64) -> u64 {
    left + right
}

#[cfg(test)]
mod tests {
    use string_interner::{StringInterner, backend::StringBackend};

    use super::*;

    #[test]
    fn it_works() {
        let mut interner = StringInterner::<StringBackend>::new();
        let lexer = Lexer::new();
        let expr_parser = ExpressionParser::new();

        let input = "console.log(\"hel\\\\lo world\", log)".to_string();

        let tokens = lexer.tokenize(&mut interner, input).unwrap();
        let expr = expr_parser
            .parse(&mut tokens.iter().peekable(), BindingPower::Default)
            .unwrap();

        println!("{:?}", expr);

        interner.iter().for_each(|(id, value)| {
            println!("{:?} => {}", id, value);
        });
    }
}
