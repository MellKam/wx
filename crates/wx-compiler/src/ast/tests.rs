use string_interner::StringInterner;
use string_interner::backend::StringBackend;

use super::parser::*;
use crate::files::Files;
use indoc::indoc;
use insta::assert_yaml_snapshot;

#[allow(unused)]
struct TestCase {
    interner: StringInterner<StringBackend>,
    files: Files,
    result: ParserResult,
}

impl<'case> TestCase {
    fn new(source: &str) -> Self {
        let mut interner = StringInterner::new();
        let mut files = Files::new();
        let file_id = files
            .add("main.wx".to_string(), source.to_string())
            .unwrap();
        let result = Parser::parse(file_id, &files.get(file_id).unwrap().source, &mut interner);

        TestCase {
            interner,
            files,
            result,
        }
    }
}

#[test]
fn should_recover_from_invalid_item() {
    let case = TestCase::new("foo func test(): unit {}");
    assert_yaml_snapshot!(case.result);
}

#[test]
fn should_recover_from_missing_function_return_type() {
    let case = TestCase::new("func test() {}");
    assert_yaml_snapshot!(case.result);
}

#[test]
fn should_recover_from_unclosed_function_block() {
    let case = TestCase::new("func test(): unit { 5 + 5 ");
    assert_yaml_snapshot!(case.result);
}

#[test]
fn should_recover_from_missing_param_separator() {
    let case = TestCase::new(indoc! {"
        func test(a: i32  b: i32): i32 { 
            a + b
        }
    "});
    assert_yaml_snapshot!(case.result);
}

#[test]
fn should_recover_from_missing_statement_separator() {
    let case = TestCase::new(indoc! {"
        func test(): i32 {
            local x = 10
            x + 5
        }
    "});
    assert_yaml_snapshot!(case.result);
}

#[test]
fn should_report_missing_local_initializer() {
    let case = TestCase::new(indoc! {"
        func test(): i32 { 
            local x: i32;
            x
        }
    "});
    assert_yaml_snapshot!(case.result);
}
