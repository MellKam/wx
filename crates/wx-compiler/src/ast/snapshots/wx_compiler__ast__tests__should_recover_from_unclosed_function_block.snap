---
source: crates/wx-compiler/src/ast/tests.rs
expression: case.result
---
ast:
  file_id: 0
  items:
    - kind:
        FunctionDefinition:
          signature:
            name:
              symbol: 1
              span:
                start: 5
                end: 9
            params:
              open:
                start: 9
                end: 10
              inner: []
              close:
                start: 10
                end: 11
            result:
              separator:
                start: 11
                end: 12
              ty:
                kind:
                  Identifier:
                    symbol: 2
                span:
                  start: 13
                  end: 17
          block:
            kind:
              Block:
                open:
                  start: 18
                  end: 19
                inner:
                  - inner:
                      Expression:
                        expr:
                          kind:
                            Binary:
                              left:
                                kind:
                                  Int:
                                    value: 5
                                span:
                                  start: 20
                                  end: 21
                              operator:
                                kind: Add
                                span:
                                  start: 22
                                  end: 23
                              right:
                                kind:
                                  Int:
                                    value: 5
                                span:
                                  start: 24
                                  end: 25
                          span:
                            start: 20
                            end: 25
                    span:
                      start: 20
                      end: 25
                    separator: ~
                close:
                  start: 26
                  end: 26
            span:
              start: 18
              end: 26
      span:
        start: 0
        end: 26
diagnostics:
  - severity: Error
    code: E0004
    message: unclosed grouping
    labels:
      - style: Primary
        file_id: 0
        range:
          start: 26
          end: 26
        message: "consider adding `}` here"
      - style: Secondary
        file_id: 0
        range:
          start: 18
          end: 19
        message: opened here
    notes: []
